schema-registry/core/src/main/java/com/hortonworks/registries/schemaregistry/SchemaVersionLifecycleManager.java

```diff
@@ -496,34 +496,29 @@ private CompatibilityResult checkCompatibility(String type,
         Collection<SchemaVersionInfo> schemaVersionInfos;
         SchemaBranchKey schemaBranchKey = new SchemaBranchKey(schemaBranchName, schemaName);
 
-        if (schemaBranchName.equals(SchemaBranch.MASTER_BRANCH)) {
-            List<QueryParam> queryParams = Collections.singletonList(new QueryParam(SchemaVersionStorable.NAME, schemaName));
-            Collection<SchemaVersionStorable> storables = storageManager.find(SchemaVersionStorable.NAME_SPACE, queryParams,
-                                                                              Collections.singletonList(OrderByField.of(SchemaVersionStorable.VERSION, true)));
-            Set<Long> schemaVersionIds = getSortedSchemaVersions(schemaBranchCache.get(SchemaBranchCache.Key.of(schemaBranchKey)))
-                    .
-                            stream()
-                    .map(schemaVersionInfo -> schemaVersionInfo.getId())
-                    .collect(Collectors.toSet());
-
-            if (storables != null && !storables.isEmpty() && schemaVersionIds != null && !schemaVersionIds.isEmpty()) {
-                schemaVersionInfos = storables
-                        .stream()
-                        .filter(schemaVersionInfo -> schemaVersionIds.contains(schemaVersionInfo.getId()))
-                        .map(SchemaVersionStorable::toSchemaVersionInfo)
-                        .collect(Collectors.toList());
-            } else {
-                schemaVersionInfos = Collections.emptyList();
-            }
-        } else {
-            schemaVersionInfos = getSortedSchemaVersions(schemaBranchCache.get(SchemaBranchCache.Key.of(schemaBranchKey)));
-            if (schemaVersionInfos == null || schemaVersionInfos.isEmpty())
-                schemaVersionInfos = Collections.emptyList();
-        }
+        schemaVersionInfos = Lists.reverse(getSortedSchemaVersions(schemaBranchCache.get(SchemaBranchCache.Key.of(schemaBranchKey))));
+        if (schemaVersionInfos == null || schemaVersionInfos.isEmpty())
+            schemaVersionInfos = Collections.emptyList();
 
         return schemaVersionInfos;
     }
 
+
+    public Collection<SchemaVersionInfo> getAllVersions(final String schemaBranchName,
+                                                        final String schemaName,
+                                                        final List<Byte> stateIds) throws SchemaNotFoundException, SchemaBranchNotFoundException {
+
+        Preconditions.checkNotNull(schemaBranchName, "Schema branch name can't be null");
+        Preconditions.checkNotNull(stateIds, "State Ids can't be null");
+
+        Set<Byte> stateIdSet = stateIds.stream().collect(Collectors.toSet());
+
+        return getAllVersions(schemaBranchName, schemaName).stream().
+                filter(schemaVersionInfo -> stateIdSet.contains(schemaVersionInfo.getStateId())).
```

Comment:
You may want to check [StorageManager#search](https://github.com/hortonworks/registry/blob/master/storage/core/src/main/java/com/hortonworks/registries/storage/StorageManager.java#L135) API for filtering values at DB layer instead of filtering on result collection. If the current API does not handle this usecase then raise github issue and this can be addressed once the raised issue is completed.

Reply:
@satishd: The values can't be filtered at database layer with current storage layer as we have to join two tables to do that. I have already raised a github issue for joins, I will wrap that up and then update this PR.

Reply:
@raju-saravanan Lets go with the current implementation. You can make respective changes in these APIs once #364 is finished.